test/random.cpp

```diff
@@ -427,3 +427,96 @@ TYPED_TEST(RandomEngineSeed, mersenneSeedUniform)
 {
     testRandomEngineSeed<TypeParam>(AF_RANDOM_ENGINE_MERSENNE_GP11213);
 }
+
+template <typename T>
+void testRandomEnginePeriod(randomEngineType type)
+{
+    if (noDoubleTests<T>()) return;
+    af::dtype ty = (af::dtype)af::dtype_traits<T>::af_type;
+
+    uint elem = 1024*1024;
+    uint steps = 4*1024;
+    af::randomEngine r(type, 0);
+
+    af::array first = af::randu(elem, ty, r);
+
+    for (int i = 0; i < steps; ++i) {
+        af::array step = af::randu(elem, ty, r);
+        bool different = !af::allTrue<bool>(first == step);
+        ASSERT_TRUE(different);
+    }
+}
+
+TYPED_TEST(RandomEngine, DISABLED_philoxRandomEnginePeriod)
+{
+    testRandomEnginePeriod<TypeParam>(AF_RANDOM_ENGINE_PHILOX_4X32_10);
+}
+
+TYPED_TEST(RandomEngine, DISABLED_threefryRandomEnginePeriod)
+{
+    testRandomEnginePeriod<TypeParam>(AF_RANDOM_ENGINE_THREEFRY_2X32_16);
+}
+
+TYPED_TEST(RandomEngine, DISABLED_mersenneRandomEnginePeriod)
+{
+    testRandomEnginePeriod<TypeParam>(AF_RANDOM_ENGINE_MERSENNE_GP11213);
+}
+
+template <typename T>
+T chi2_statistic(array input, array expected) {
+    expected *= af::sum<T>(input) / af::sum<T>(expected);
+    array diff = input - expected;
+    return af::sum<T>((diff * diff) / expected);
+}
+
+template <typename T>
+void testRandomEngineUniformChi2(randomEngineType type)
+{
+    if (noDoubleTests<T>()) return;
+    af::dtype ty = (af::dtype)af::dtype_traits<T>::af_type;
+
+    int elem = 256*1024*1024;
+    int steps = 32;
+    int bins = 100;
+
+    array total_hist = af::constant(0.0, bins, ty);
+    array expected = af::constant(1.0/bins, bins, ty);
+
+    af::randomEngine r(type, 0);
+
+    // R> qchisq(c(5e-6, 1 - 5e-6), 99)
+    // [1]  48.68125 173.87456
+    T lower = 48.68125;
+    T upper = 173.87456;
+
+    bool prev_step = true;
+    bool prev_total = true;
+    for (int i = 0; i < steps; ++i) {
+        array step_hist = af::histogram(af::randu(elem, ty, r), bins, 0.0, 1.0);
+        T step_chi2 = chi2_statistic<T>(step_hist, expected);
+        bool step = step_chi2 > lower && step_chi2 < upper;
+        ASSERT_TRUE(step || prev_step);
```

Comment:
Would these `EXPECT_*` be instead or in addition to the existing `ASSERT_TRUE`?

Reply:
Replacing the assertions covering `step` and `prev_step` with assertions covering only `step` might produce false positives. When testing RNGs some failures are to be expected and it is the question where to draw the line. However, the output from these assertions is indeed much nicer. I see two possibilities:

* Only test the `EXPECT` when `prev_step` is already `false`:

        if (!prev_step) {      
            EXPECT_GT(step_chi2, lower) << "at step: " << i;
            EXPECT_LT(step_chi2, upper) << "at step: " << i;
        }
* Accept a larger range of chi² values *and* do more steps.

Any preferences? At the moment I would prefer the first option, since it does not require retesting the failure condition.

Reply:
Yeah, I think the first option is the way to go.